src/middleware/logger.ts

```diff
@@ -39,3 +39,94 @@ export class DefaultLogger implements Logger {
     console.debug(...args);
   }
 }
+
+/**
+ * LogBuffer hold the args that will be passed on to the actual
+ * logger when attempting to flush.
+ */
+export class LogBuffer {
+  public readonly level: string;
+  public readonly args: LogArg[];
+
+  constructor(level: string, ...args: LogArg[]) {
+    this.level = level;
+    this.args = args;
+  }
+}
+
+/**
+ * ProxyLogger attempts to temporarily hold user's logs
+ * during the function run, and attempt to flush it through
+ * the provided logger when the function returns.
+ *
+ * The expected usage of this class is,
+ * 1. store the log attempt with its arguments and level
+ * 2. if function hits a step that has been memorized, clear the _buffer
+ * 3. on function return, flush all buffers
+ *
+ * And it should be invisible to the user as much as possible.
+ */
+export class ProxyLogger implements Logger {
+  private readonly _logger: Logger;
+  private _buffer: LogBuffer[] = [];
+
+  constructor(logger: Logger) {
+    this._logger = logger;
+  }
+
+  info(...args: LogArg[]) {
+    this._buffer.push(new LogBuffer("info", ...args));
+  }
+
+  warn(...args: LogArg[]) {
+    this._buffer.push(new LogBuffer("warn", ...args));
+  }
+
+  error(...args: LogArg[]) {
+    this._buffer.push(new LogBuffer("error", ...args));
+  }
+
+  debug(...args: LogArg[]) {
+    this._buffer.push(new LogBuffer("debug", ...args));
+  }
+
+  reset() {
+    this._buffer = [];
+  }
+
+  async flush() {
+    if (this.bufSize() === 0) return;
+
+    // eslint-disable-next-line @typescript-eslint/require-await
+    const deliveries = this._buffer.map(async (log) => {
+      const args = log.args;
+      const method = log["level"] as keyof Logger;
+      return this._logger[method](...args);
+    });
+    // NOTE: timestamp will likely not be linear as expected.
+    await Promise.all(deliveries);
+    this.reset();
+
+    // Allow 1s for the provided logger to handle flushing since the ones that do
+    // flushing usually has some kind of timeout of up to 1s.
+    //
+    // TODO:
+    // This should only happen when using a serverless environment because it's very
+    // costly from the compute perspective.
+    // server runtimes should just let the logger do their thing since most of them
+    // should have already figured what to do in those environments, be it threading or
+    // something else.
+    if (this._logger.constructor.name !== DefaultLogger.name) {
+      await new Promise((resolve) => {
+        setTimeout(() => resolve(null), 1000);
```

Comment:
I'm confused by this. What are we waiting 1 second for? Haven't we already awaited the deliveries?

Reply:
Because based on how the user setup their logger, the implementation details within the provided logger could have timeouts or sleep, when they actually handle the delivery.
Just calling `_logger.info` doesn't actually mean the provided logger will attempt delivery right away.

Some examples,

*Pino*
https://github.com/pinojs/pino/blob/master/docs/transports.md#asynchronous-startup
https://github.com/pinojs/pino/blob/171d8056a2a66c42e4bd2e13b5936b6a652af830/lib/transport.js#L53

*Winston*
https://github.com/winstonjs/winston/blob/f8320c1e3bb7cc3544b1385da92cc625e2b3b0ac/lib/winston/transports/http.js#L199-L203

It's not consistent, but 1s is a guess that could balance enough time to flush vs hit to throughput.